### PR TITLE
Material: Add alpha to coverage define, add clipping & alpha test anti aliasing

### PR DESCRIPTION
**Pull request by gkjohnson**
Related issue: --

**Description**

This PR is a draft to add a `#define` for `ALPHA_TO_COVERAGE` to materials when `material.alphaToCoverage = true` and updates materials clipping planes to take advantage of the option to antialias clipped edges. I don't expect that the `alphaToCoverage` field will be toggled frequently so hopefully adding a new define in this case is okay. The define can eventually be used for other applications like smoothing aliased edges when `alphaTest` is true. If the feature is agreeable I can go ahead and make the rest of the needed changes.

A couple things to be aware of:

- clipping with alpha to coverage = true requires a few more shader operations per clip plane.
- clipping with alpha to coverage = true requires the derivatives extension (which shouldn't be a problem in WebGL2).
- transparent objects will still have an aliased clip edge. This _could_ be addressed but would require a new define for `transparent = true` objects which is probably not worth it. Or we could just use the new clipping method code always which will work even if alpha to coverage = false but be slightly more expensive.
- all materials will have to be adjusted so the diffuseColor definition comes before the the clipping planes shader chunk.

Some updated demos with an "alphaToCoverage" GUI toggle:

[webgl_clipping](https://raw.githack.com/gkjohnson/three.js/alpha-to-coverage-define-demo/examples/webgl_clipping.html)

[webgl_clipping_intersection](https://raw.githack.com/gkjohnson/three.js/alpha-to-coverage-define-demo/examples/webgl_clipping_intersection.html)

[webgl_clipping_stencil](https://raw.githack.com/gkjohnson/three.js/alpha-to-coverage-define-demo/examples/webgl_clipping_stencil.html)

And some images:

| Before | After |
|---|---|
| image | image |
| image | image |

### CODE REVIEW

**Comment by WestLangley** [comment]
A very informative post

https://bgolus.medium.com/anti-aliased-alpha-test-the-esoteric-alpha-to-coverage-8b177335ae4f

**Comment by gkjohnson** [comment]
> The define can eventually be used for other applications like smoothing aliased edges when alphaTest is true.

Here's an example of how the define can be used to smooth alpha test materials using textures with smooth gradients. The alphaToCoverage setting will already help smooth out alpha test textures with a stark alpha contrast but as mentioned in the article @WestLangley posted it can still cause issues with mip maps and linear texture interpolation:

| Settings | Screenshot |
|---|---|
| transparent = true | image |
| alphaTest = 0.5 | image |
| alphaTest = 0.5, alphaToCoverage = true (dev branch) | image |
| alphaTest = 0.5, alphaToCoverage = true (with A2C define) | image |

And here's a zoomed in shot of the difference in edge quality (left is dev, right is new alphaToCoverage + alphaTest):

image

**Comment by gkjohnson** [comment]
@mrdoob is this interesting to you? I'm happy to finish out adding support for clip plane edges and alpha test textures if so.

[This Oculus VR guide](https://developer.oculus.com/blog/common-rendering-mistakes-how-to-find-them-and-how-to-fix-them/?locale=es_ES) recommends using alpha to coverage for alpha cutout materials. And [here's an article](https://webcache.googleusercontent.com/search?q=cache:40SiBiBSi-gJ:https://www.pcgameshardware.de/Street-Fighter-4-Spiel-35526/News/Street-Fighter-IV-PC-explained-in-detail-685997/+&cd=5&hl=en&ct=clnk&gl=us) (cached version linked because the main site has a cookie pop up that's difficult to navigate) that discusses Street Fighter 4 using alpha to coverage for fences and leaves:

> Alpha to Coverage
> It may be a little difficult to spot but if your video card is compatible, alpha to coverage is applied. Objects that use Alpha test will look different when MSAA is in use. This is mainly applied to objects such as fences and leaves.

**Comment by gkjohnson** [comment]
I think this should be ready, now. It includes the ability to use alpha to coverage to smooth edges when using material clipping planes and alpha test for alpha maps. I've updated a couple examples, as well, to show its capabilities.

You can see the differences here:

image

image

**Comment by mrdoob** [comment]
Very nice! Is there any case in which we wouldn't want to have this always enabled?

**Comment by wmcmurray** [comment]
> Very nice! Is there any case in which we wouldn't want to have this always enabled?

Also when a user **want** to display very sharply pixelated transparent texture (low resolution + alphaTest + NearestFilter), I suppose this feature would interfere with the ability to do that ? 🤔

**Comment by mrdoob** [comment]
> Unfortunately when it's enabled for partially transparent objects you get this kind of banding where there should be smooth gradients since the opacity is mapped to one of the 4 (8?) MSAA coverage levels.

How about when `transparent === false` and `alphaTest > 0 && alphaTest < 1`?

**Comment by gkjohnson** [comment]
> How about when `transparent === false` and `alphaTest > 0 && alphaTest < 1`?

With this PR `Material.alphaToCoverage` now enables the the `SAMPLE_ALPHA_TO_COVERAGE` gl parameter _and_ adds a `#define` to the shaders so shader features like clipping and alphaTest can take advantage of it. AlphaToCoverage isn't only relevant in those cases, though. Users writing custom shaders may want to enable A2C without using alphatest.

And it can actually still be worthwhile to enable A2C when using an alpha map and _not_ use alpha test. Some of the articles I referenced in [this comment](https://github.com/mrdoob/three.js/issues/20522#issuecomment-714178497) discuss how A2C can be used to alleviate some of the apparent volume loss when using alpha test with texture mipmaps, though there may need to be other changes needed to make this "just work".

That's all to say that A2C has a variety of uses beyond just the ones added here so I don't think this should be an automatically toggled.

**Comment by mrdoob** [comment]
Yes, sorry. I was basically reconsidering `material.alphaToCoverage`.

1. Does the dev/user needs to be aware of this or can we just take care of it for them?
2. Is that the best name for it? Could it be something like `material.alphaTestSmooth = true|false`?
3. Is this WebGL only or does it apply to WebGPU too? Does the name `alphaToCoverage` makes sense in WebGPU?

Mainly looking for API suggestions that make this more intuitive for new devs.

**Comment by gkjohnson** [comment]
> Does the dev/user needs to be aware of this or can we just take care of it for them?

I don't know. I don't think we can but perhaps I'm just overly detail oriented. I actually removed the "alpha to coverage" toggle for the `webgl_clipping_stencil` example because it actually caused more artifacts at the seams of the clip planes no matter how I applied the alpha at the edges. It either often created a gap at the seam between the edges or created a doubling of edges where the planes intersected. I feel like there are cases where you'd want to disable this.

> Is that the best name for it? Could it be something like `material.alphaTestSmooth = true|false`?

I hesitate to use names that obscure what's actually happening. "Alpha To Coverage" is a searchable term that gives information on what's actually happening at the GPU level and I think that's a good thing. Something like "alphaTestSmooth" also implies this only works in a single use case -- it doesn't. End users can write shaders with the alpha to coverage behavior in mind to get smooth edges with procedural fragment shaders and has benefits even when alphaTest isn't bein used.

> Is this WebGL only or does it apply to WebGPU too? Does the name alphaToCoverage makes sense in WebGPU?

This API feature is referred to as "alpha to coverage" in [Vulkan](https://www.khronos.org/registry/vulkan/specs/1.3-extensions/man/html/VkPipelineMultisampleStateCreateInfo.html), [DirectX](https://docs.microsoft.com/en-us/windows/win32/direct3d11/d3d10-graphics-programming-guide-blend-state#alpha-to-coverage), and [WebGPU](https://www.w3.org/TR/webgpu/#dom-gpumultisamplestate-alphatocoverageenabled).

**Comment by gkjohnson** [comment]
@mrdoob I see it hasn't been added to any milestone -- is this blocked because it can't be implicitly enabled? 

This PR just enhances the existing Material.alphaToCoverage flag and doesn't add any new APIs so if we wanted to reevaluate how alphaToCoverage is toggled I think that could happen in a future PR.

**Comment by sdarpeng** [comment]
Why alphaToCoverage doesn't work on r145 on my computer ?  I copied the code by @gkjohnson local , but alpha-test antialiasing still failed.

**Comment by sdarpeng** [comment]
And the documentation about alphaToCoverage is still wrong for it's value should be boolean not float.

**Comment by gkjohnson** [comment]
I'm not sure what's up the "webgl_shadowmap_pointlight" screenshot here - can someone else give regenerating it a try?

File: src/renderers/shaders/ShaderLib/meshbasic.glsl.js [hunk 1]
```diff
@@ -68,9 +68,8 @@ uniform float opacity;
 
 void main() {
 
-	#include <clipping_planes_fragment>
-
 	vec4 diffuseColor = vec4( diffuse, opacity );
+	#include <clipping_planes_fragment>
```
Comment (alberto-taiuti-skydio) [inline src/renderers/shaders/ShaderLib/meshbasic.glsl.js:72]: I'm curious, what's the reason for moving the call to the clipping code below the calculation for the diffuse color, in many of the shaders?

Comment (gkjohnson) [reply src/renderers/shaders/ShaderLib/meshbasic.glsl.js:72]: The clipping fragment code block now [multiples into the diffuse color alpha](https://github.com/mrdoob/three.js/pull/22172/files#diff-510604884d1277c044a3c6f6fc1994ae2baaedd6e7cc74a24df80017074dd52cR43) to perform the alpha test AA.

Comment (alberto-taiuti-skydio) [reply src/renderers/shaders/ShaderLib/meshbasic.glsl.js:72]: Ah, makes sense. Thanks for explaning

**Comment by gkjohnson** [comment]
@Mugen87 @mrdoob is there any way to see the output screenshot from CI to compare with the uploaded screenshots to see what the differences might be?

**Comment by Mugen87** [comment]
I remember a tool that did show the diff as red/pink pixels or something similar. However, I don't know anymore how it was used and whether it was ever integrated in the CI.

**Comment by LeviPesin** [comment]
> However, I don't know anymore how it was used and whether it was ever integrated in the CI.

It is an essential part of the E2E test -- the difference screenshots are autogenerated, so you can just download the artefact from the run and see them there.

**Comment by gkjohnson** [comment]
> It is an essential part of the E2E test -- the difference screenshots are autogenerated, so you can just download the artefact from the run and see them there.

Thanks - finding the artifacts in the Github UI is non trivial.

---

I've reverted example changes since it's possible there are small differences in how the CI and my local machine are performing A2C sampling:

| Expected | Actual | Diff |
|---|---|---|
| webgl_shadowmap_pointlight-expected | webgl_shadowmap_pointlight-actual | webgl_shadowmap_pointlight-diff |

You can see they're both anti-aliased at the stipe edgues but still look different.

The last issue is with `webgl_nodes_materials_instance_uniform` which I'm confused by beacause I don't _think_ node materials touches anything in this PR? cc @sunag, any ideas? It looks like the env map is being sampled in sampled in screenspace for some reason in the "after" screenshot:

| before | after |
|---|---|
| webgl_nodes_materials_instance_uniform-expected | webgl_nodes_materials_instance_uniform-actual |

**Comment by sunag** [comment]
> The last issue is with webgl_nodes_materials_instance_uniform which I'm confused by beacause I don't think node materials touches anything in this PR? cc @sunag, any ideas? It looks like the env map is being sampled in sampled in screenspace for some reason in the "after" screenshot:

`WebGLNodeBuilder` still works with simple code replacement, so the order of the compiled nodes with those of the replaced code must be the same, I pushed a small commit to correct this.

**Comment by gkjohnson** [comment]
@sunag thank you!

Well this is ready to merge, again, if we could.

**Review by Mugen87 — APPROVED** [review]

**Pull request merged**